From 8353df8d8a56e9862d0d3602ff73aa2a04b08fb4 Mon Sep 17 00:00:00 2001
From: Phil Steitz <psteitz@apache.org>
Date: Tue, 29 Jan 2008 02:29:00 +0000
Subject: [PATCH] Fixed error in SharedPoolDataSource causing incorrect
 passwords to be stored under certain conditions. JIRA: DBCP-245 Reported and
 patched by Michael Drechsel

git-svn-id: https://svn.apache.org/repos/asf/commons/proper/dbcp/trunk@616142 13f79535-47bb-0310-9956-ffa450edef68
---
 .../commons/dbcp/datasources/SharedPoolDataSource.java     | 7 +++++++
 .../commons/dbcp/datasources/TestSharedPoolDataSource.java | 6 +++++-
 xdocs/changes.xml                                          | 4 ++++
 3 files changed, 16 insertions(+), 1 deletion(-)

diff --git a/src/java/org/apache/commons/dbcp/datasources/SharedPoolDataSource.java b/src/java/org/apache/commons/dbcp/datasources/SharedPoolDataSource.java
index 99b7af31e4..b343ce2889 100644
--- a/src/java/org/apache/commons/dbcp/datasources/SharedPoolDataSource.java
+++ b/src/java/org/apache/commons/dbcp/datasources/SharedPoolDataSource.java
@@ -167,6 +167,13 @@ public int getNumIdle() {
             info = (PooledConnectionAndInfo) pool
                 .borrowObject(getUserPassKey(username, password));
         }
+        catch (SQLException ex) {  // Remove bad UserPassKey
+            if ((userKeys != null) && (userKeys.containsKey(username))) {
+                userKeys.remove(username);
+            }
+            throw new SQLNestedException(
+                "Could not retrieve connection info from pool", ex);
+        }
         catch (Exception e) {
             throw new SQLNestedException(
                 "Could not retrieve connection info from pool", e);
diff --git a/src/test/org/apache/commons/dbcp/datasources/TestSharedPoolDataSource.java b/src/test/org/apache/commons/dbcp/datasources/TestSharedPoolDataSource.java
index 586dbef225..fe45aa8423 100644
--- a/src/test/org/apache/commons/dbcp/datasources/TestSharedPoolDataSource.java
+++ b/src/test/org/apache/commons/dbcp/datasources/TestSharedPoolDataSource.java
@@ -78,9 +78,14 @@ public void testBackPointers() throws Exception {
      * Switching 'u1 -> 'u2' and 'p1' -> 'p2' will
      * exhibit the bug detailed in 
      * http://issues.apache.org/bugzilla/show_bug.cgi?id=18905
+     * 
+     * Starting with a successful connection, then incorrect password,
+     * then correct password for same user illustrates
+     * JIRA: DBCP-245
      */
     public void testIncorrectPassword() throws Exception 
     {
+        ds.getConnection("u2", "p2").close();
         try {
             // Use bad password
             ds.getConnection("u1", "zlsafjk").close();
@@ -520,6 +525,5 @@ public void testPoolPreparedStatements() throws Exception {
         assertTrue(l3HashCode == l4HashCode);
         conn.close();
         conn = null;
-
     }
 }
diff --git a/xdocs/changes.xml b/xdocs/changes.xml
index e8f0043f3e..7b52ccfc09 100644
--- a/xdocs/changes.xml
+++ b/xdocs/changes.xml
@@ -101,6 +101,10 @@ The <action> type attribute can be add,update,fix,remove.
         removeConnectionEventListener from connection event handlers to
         eliminate potential for ConcurrentModificationExceptions.
       </action>
+      <action dev="psteitz" type="fix" issue="DBCP-245" due-to="Michael Drechsel">
+        Fixed error in SharedPoolDataSource causing incorrect passwords to be 
+        stored under certain conditions.
+      </action>   
     </release>
     <release version="1.2.2" date="2007-04-04"
       description="This is a maintenance release containing bug fixes
